### PR TITLE
Now solveset returning solutions in easy form.

### LINKED ISSUE #9824: solveset fails on sin(x)**2 - 2*sin(x) + 1
`solveset` raises `NotImplementedError` on the above equation.

However, a similar variation is solved :
`>>> solveset(cos(x)**2 - 2*cos(x) + 1, x, domain=S.Reals)`
`ImageSet(Lambda(_n, 2*_n*pi), Integers())`

### LINKED ISSUE #10426: Can't solve sin(x + a) - sin(x) (for a)
```
In [9]: solveset(sin(x + a) - sin(x), a)
Out[9]: {a | a ∊ ℂ ∧ -sin(x) + sin(a + x) = 0}

In [10]: solveset(sin(x + a) - sin(x), a, domain=S.Reals)
Out[10]:
⎧              ⎛  ⎛   ⅈ⋅(-a - x)    ⅈ⋅(a + x)⎞  ⅈ⋅x    2⋅ⅈ⋅x    ⎞  -ⅈ⋅x    ⎫
⎪            ⅈ⋅⎝- ⎝- ℯ           + ℯ         ⎠⋅ℯ    + ℯ      - 1⎠⋅ℯ        ⎪
⎨a | a ∊ ℝ ∧ ────────────────────────────────────────────────────────── = 0⎬
⎪                                        2                                 ⎪
⎩                                                                          ⎭

In [11]: solve(sin(x + a) - sin(x), a)
Out[11]: [-x + asin(sin(x)), -x - asin(sin(x)) + π]

```

The solution should be `a=2*pi*n`. I believe the first solution from solve is equivalent to that.

### PR DESCRIPTION
**Pull request by Shekharrajak**
Tried to fix https://github.com/sympy/sympy/issues/10426

This may able to give solutions in reduced and known easy form in many cases.

```
>>> print solveset_real(cos(x), x)
ImageSet(Lambda(_n, _n*pi + pi/2), Integers())
>>> print solveset_real(sin(x), x)
ImageSet(Lambda(_n, _n*pi), Integers())
>>> print solveset_real(tan(x), x)
ImageSet(Lambda(_n, _n*pi), Integers()) \ ImageSet(Lambda(_n, _n*pi + pi/2), Integers())
>>> solveset(sin(x + a) - sin(x), a, domain=S.Reals) 
 ImageSet(Lambda(n, 2*n*pi), S.Integers)
```

Edit 1 : 
Also fixes https://github.com/sympy/sympy/issues/9824

### CODE REVIEW

**Comment by Shekharrajak** [comment]
Github is not updating my master branch and giving branch conflict:(

File: sympy/solvers/tests/test_solveset.py [hunk 1]
```diff
@@ -709,22 +709,10 @@ def test_solveset_complex_tan():
...
-    assert solveset_real(cos(x), x) == \
-        Union(imageset(Lambda(n, 2*pi*n - pi/2), S.Integers),
-              imageset(Lambda(n, 2*pi*n + pi/2), S.Integers))
-
-    assert solveset_real(sin(x) + cos(x), x) == \
-        Union(imageset(Lambda(n, 2*n*pi - pi/4), S.Integers),
-              imageset(Lambda(n, 2*n*pi + 3*pi/4), S.Integers))
-
-    assert solveset_real(sin(x)**2 + cos(x)**2, x) == S.EmptySet
+    assert solveset_real(sin(x)**2 + cos(x)**2, x) == S.EmptyS
```
Comment (asmeurer) [inline sympy/solvers/tests/test_solveset.py]: Why did you delete these tests?

File: sympy/solvers/solveset.py [hunk 2]
```diff
@@ -539,23 +539,7 @@ def _solve_real_trig(f, symbol):
...
-
-    solns = solveset_complex(g, y) - solveset_complex(h, y)
-
-    if isinstance(solns, FiniteSet):
-        return Union(*[invert_complex(exp(I*symbol), s, symbol)[1]
-                       for s in solns])
-    elif solns is S.EmptySet:
-        return S.EmptySet
-    else:
-        return ConditionSet(symbol, Eq(f, 0), S.Reals)
```
Comment (asmeurer) [inline sympy/solvers/solveset.py]: Why did you delete this code? Assumedly it was here for a reason. You should explain more in your commit messages.

Comment (aktech) [inline sympy/solvers/solveset.py]: I don't see a reason for this.

File: sympy/solvers/tests/test_solveset.py (hunk 1 above)
Comment (Shekharrajak) [inline sympy/solvers/tests/test_solveset.py]: These are already [here](https://github.com/Shekharrajak/sympy/blob/master/sympy/solvers/tests/test_solveset.py#L749).
Now `solveset` returns desired output for these.which was previously @XFAIL. @asmeurer

File: sympy/solvers/solveset.py (hunk 2 above)
Comment (Shekharrajak) [inline sympy/solvers/solveset.py]: [This line ](https://github.com/Shekharrajak/sympy/blob/master/sympy/solvers/solveset.py#L546) substituting `y` in place of `I*x` , it is okay for simple `exp` equations , but when we get some complicated equations like `exp(I*x/2)` ,then it is not able to substitute `y`.Most of the time because of this, [ here if condition](https://github.com/Shekharrajak/sympy/blob/master/sympy/solvers/solveset.py#L547) becomes `true` and returns `conditionSet`.So it is not able to solve many equation which can be solved easily.
If I use ` return solveset(f,symbol)` then now `f` is in `exp` form which is `complex` type and `solveset` is able to solve complicated `exp` equations.
So it gives correct output in reduced form which is desired.
@asmeurer  @aktech

Comment (Shekharrajak) [inline sympy/solvers/solveset.py]: Even though if `y` is substituted then also [this line](https://github.com/Shekharrajak/sympy/blob/master/sympy/solvers/tests/test_solveset.py#L749) not always able to give simple form.So there are some chance that solution will be [`FiniteSet`](https://github.com/sympy/sympy/pull/10482#discussion-diff-51149813L552).
Hence many times it end up [here](https://github.com/sympy/sympy/pull/10482#discussion-diff-51149813L557) with `conditionSet`, that we don't want.

Comment (asmeurer) [inline sympy/solvers/solveset.py]: This is the sort of thing which is good to write in the commit message.

Comment (Shekharrajak) [inline sympy/solvers/solveset.py]: Yeah , actually I am struggling with branch conflict,don't know why my master branch is not updated the `test_solveset.py` file.I am trying to add main test-case for the issue.
Next time I will keep this in mind.

**Comment by smichr** [comment]
I would like to see #10448 be committed first, if possible.

File: sympy/solvers/solveset.py (hunk 2 above)
Comment (smichr) [inline sympy/solvers/solveset.py]: This change seems right except that `solveset_real` should be used if we are in the `_solve_real_trig` function. It should not matter that the argument of `exp` might be imaginary (i.e. `exp(I*x)`); the solutions for `x` will be real.

**Comment by smichr** [comment]
The tests that you have deleted should not be deleted. Was that an accident?

Regarding #10448 -- I see now that it won't matter too much if this or that is done first.

**Comment by Shekharrajak** [comment]
[Previous testcase](https://github.com/sympy/sympy/pull/10482/files#diff-85baa04bbf4e1dfd9128782738e45424L712)  have different result from [this](https://github.com/sympy/sympy/pull/10482/files#diff-85baa04bbf4e1dfd9128782738e45424L751).

File: sympy/solvers/solveset.py (hunk 2 above)
Comment (Shekharrajak) [inline sympy/solvers/solveset.py]: In [return statement](https://github.com/sympy/sympy/pull/10482/files#diff-eec0422923e8f100745c015cd8fdd6cfR542) we are passing `f` which is now in `exp` form and `domain= complexes` so `solveset` take this to [here](https://github.com/Shekharrajak/sympy/blob/master/sympy/solvers/solveset.py#L968) `solveset_complex`.I just wanted to make sure that if there is more complicated/nested `exp` form then it should able to handle that automatically.Isn't it ?

**Comment by Shekharrajak** [comment]
https://github.com/sympy/sympy/pull/10448  is not changing anything in `_solve_real_trig` so I hope there won't be any branch conflict.
 @smichr I think you can handle https://github.com/sympy/sympy/issues/10217 
with some changes in [this block of code](https://github.com/sympy/sympy/pull/10448/files#diff-eec0422923e8f100745c015cd8fdd6cfR584).Just a problem is `rh_s` is in `imageSet`.

File: sympy/solvers/solveset.py (hunk 2 above)
Comment (smichr) [inline sympy/solvers/solveset.py]: Not as I understand it. Though it is not enforced well, it is my understanding that the purpose of the `*_real` functions is to use the fact that the domain is real to select the correct/most appropriate solvers for the solving process. By allowing `f` to be solved in the complex domain you will allow this routine to send back complex solutions which is not desired. It's not a matter of "complication" it's a matter of what solution from the arbitrarily complex expression should be returned. 

However, seeing that there is not `_complex` trig routine and that the former function did the same, it seems that what you have done is no worse than what was present before.

File: sympy/solvers/tests/test_solveset.py [hunk 3]
```diff
@@ -709,21 +709,9 @@ def test_solveset_complex_tan():
 
 def test_solve_trig():
     from sympy.abc import n
-    assert solveset_real(sin(x), x) == \
```
Comment (smichr) [inline sympy/solvers/tests/test_solveset.py]: Why are these tests deleted?

Comment (Shekharrajak) [inline sympy/solvers/tests/test_solveset.py]: I [ commented ](https://github.com/sympy/sympy/pull/10482/files#diff-85baa04bbf4e1dfd9128782738e45424R1059) it

File: sympy/solvers/solveset.py (hunk 2 above)
Comment (Shekharrajak) [inline sympy/solvers/solveset.py]: If modified  [this line ](https://github.com/sympy/sympy/pull/10482/files#diff-eec0422923e8f100745c015cd8fdd6cfR542) with `return solveset_real(f,symbol)`

Then it is giving solution in ` exp` form ,that is not we want , right?
```
>>> print solveset(sin(x + a) - sin(x), a, domain=S.Reals)
ConditionSet(a, Eq(-exp(2*I*x) - exp(I*x)*exp(I*(-a - x)) + exp(I*x)*exp(I*(a + x)) + 1, 0), (-oo, oo))
>>> print solveset(sin(x + a) - sin(x), a)
ConditionSet(a, Eq(-sin(x) + sin(a + x), 0), Complexes(Lambda((_x, _y), _x + _y*I), (-oo, oo) x (-oo, oo)))

```
we can `return solveset_complex(f,symbol))` for solution 
```
>>> solveset(sin(x + a) - sin(x), a)
{a | a ∊ ℂ ∧ -sin(x) + sin(a + x) = 0}
>>> solveset(sin(x + a) - sin(x), a, domain=S.Reals)
{2⋅n⋅π | n ∊ ℤ}
```
which is same as what `return solveset(f,symbol)` gives.

File: sympy/solvers/tests/test_solveset.py (hunk 3 above)
Comment (smichr) [inline sympy/solvers/tests/test_solveset.py]: You can replace this (and others similarily) with `assert solveset(sin(x), x, S.Reals) == imageset(Lambda(n, pi*n), S.Integers)`

**Comment by hargup** [comment]
@Shekharrajak There are merge conflicts in this PR.

**Comment by Shekharrajak** [comment]
Closing this PR, I have extended these changes here : https://github.com/sympy/sympy/issues/10733
